quadriga/src/main/java/edu/asu/diging/quadriga/core/citesphere/impl/CitesphereConnectorImpl.java

```diff
@@ -74,31 +85,89 @@ public CitesphereConnectorImpl() {
-                return restTemplate.exchange(appUrl, HttpMethod.GET, entity,
+                return restTemplate.exchange(appUrl, HttpMethod.GET, generateCheckTokenEntity(),
                         new ParameterizedTypeReference<List<CitesphereAppInfo>>() {
                         }).getBody();
             }
             throw ex;
         }
     }
+    
+    /**
+     * This method accepts a token from Vogon and sends it to Citesphere's "check_token"
+     * end point for token validation using an access token
+     * This access token is used by quadriga to talk to citesphere
+     * 
+     * If Citesphere responds with unauthorized HTTP exception, we try to re-generate
+     * the access token, in case it is expired.
+     * 
+     * @param token is the one we receive from Vogon to check the validity
+     * @return the token response received from citesphere
+     * @throws BadCredentialsException if token/access token is invalid
+     * @throws OAuthException if request is unauthorized because of the token/access token
+     */
+    @Override
+    public TokenInfo getTokenInfo(String token) throws BadCredentialsException, OAuthException {
+        
+
+        String checkTokenUrl = citesphereBaseUrl + citesphereCheckTokenEndpoint + "?token=" + token;
+        TokenInfo tokenInfo = null;
+        
+        try {
+            tokenInfo = restTemplate.postForObject(checkTokenUrl, generateCheckTokenEntity(), TokenInfo.class, new Object[] {});
+        } catch (HttpClientErrorException e1) {
+            
+            if (e1.getStatusCode() == HttpStatus.UNAUTHORIZED) {
+                
+                // If we get an unauthorized exception, the accessToken could be expired
+                // so we generate a new access token and we again try to call checkToken URL
+                currentAccessToken = getAccessToken();
+                try {
+                    tokenInfo = restTemplate.postForObject(checkTokenUrl, generateCheckTokenEntity(), TokenInfo.class, new Object[] {});
+                }  catch(HttpClientErrorException e2) {
+
+                    //  If we again get an unauthorized exception, we will just throw an OAuthException
+                    if (e2.getStatusCode() == HttpStatus.UNAUTHORIZED) {
+                        throw new OAuthException();
+                    }
+                    throw new BadCredentialsException("Token is invalid for app.", e1);
+                }
+            } else {
+            
+                // If it is some other kind of exception, we throw a BadCredentialsException
+                throw new BadCredentialsException("Token is invalid for app.", e1);
+            }
+        }
+
+        return tokenInfo;
+    }
+    
+    
+    /**
+     * Here we generate an HTTP entity based on the accessToken
+     * If accessToken is absent, we ask the getAccessToken method to generate one
+     * 
+     * @return an HTTP Entity
+     */
+    private HttpEntity<String> generateCheckTokenEntity() {
```

Comment:
this method is independent of the check token functionality, is it? it just generates an httpEntity with an access ttoken.